src/Evolution/Systems/Burgers/FiniteDifference/BoundaryConditionGhostData.hpp

```diff
@@ -0,0 +1,239 @@
+#include "Domain/Structure/DirectionMap.hpp"
+#include "Domain/Structure/Element.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Domain/Tags.hpp"
+#include "Domain/TagsTimeDependent.hpp"
+#include "Evolution/BoundaryConditions/Type.hpp"
+#include "Evolution/DgSubcell/Tags/Mesh.hpp"
+#include "Evolution/DgSubcell/Tags/NeighborData.hpp"
+#include "Evolution/DiscontinuousGalerkin/NormalVectorTags.hpp"
+#include "Evolution/Systems/Burgers/BoundaryConditions/BoundaryCondition.hpp"
+#include "Evolution/Systems/Burgers/FiniteDifference/Reconstructor.hpp"
+#include "Evolution/Systems/Burgers/System.hpp"
+#include "Evolution/Systems/Burgers/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Parallel/Tags/Metavariables.hpp"
+#include "Utilities/CallWithDynamicType.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Burgers::fd {
+/*!
+ * \brief Computes finite difference ghost data for external boundary
+ * conditions.
+ *
+ * If the element is at the external boundary, computes FD ghost data with a
+ * given boundary condition and store it into neighbor data with {direction,
+ * ElementId::external_boundary_id()} as the mortar_id key.
+ *
+ * \note Subcell needs to be enabled for boundary elements (see
+ * Burgers::subcell::TimeDerivative). Otherwise this function would be never
+ * called.
+ *
+ */
+struct BoundaryConditionGhostData {
+  template <typename DbTagsList>
+  static void apply(const gsl::not_null<db::DataBox<DbTagsList>*> box,
+                    const Element<1>& element,
+                    const Burgers::fd::Reconstructor& reconstructor);
+
+ private:
+  template <typename FdBoundaryConditionHelper, typename DbTagsList,
+            typename... FdBoundaryConditionArgsTags>
+  // A helper function for calling fd_ghost() of BoundaryCondition subclasses
+  static void apply_subcell_boundary_condition_impl(
+      FdBoundaryConditionHelper& fd_boundary_condition_helper,
+      const gsl::not_null<db::DataBox<DbTagsList>*>& box,
+      tmpl::list<FdBoundaryConditionArgsTags...>) {
+    return fd_boundary_condition_helper(
+        db::get<FdBoundaryConditionArgsTags>(*box)...);
+  }
+};
+
+template <typename DbTagsList>
+void BoundaryConditionGhostData::apply(
+    const gsl::not_null<db::DataBox<DbTagsList>*> box,
+    const Element<1>& element,
+    const Burgers::fd::Reconstructor& reconstructor) {
+  const ::Domain<1>& domain = db::get<domain::Tags::Domain<1>>(*box);
+  const auto& external_boundary_condition =
+      domain.blocks()[element.id().block_id()].external_boundary_conditions();
+
+  const Mesh<1> subcell_mesh =
+      db::get<evolution::dg::subcell::Tags::Mesh<1>>(*box);
+
+  const size_t ghost_zone_size{reconstructor.ghost_zone_size()};
+
+#ifdef SPECTRE_DEBUG
+  bool element_at_external_boundary = 0;
+#endif
+
+  for (const auto& direction : Direction<1>::all_directions()) {
```

Comment:
Can you not combine the `for` and `if` into `for (const auto& direction : element.external_boundaries()) {`? It seems like the `ASSERT` on line 97 is really just checking that the element was constructed in such a way that the external boundaries are different from the neighbors, or what am I missing?

Reply:
Right, I don't think you want to get rid of a check like that, but isn't `ASSERT(not element.external_boundaries().empty(), "");` enough? That is, you require there be some external boundaries?

Reply:
Oops. squashed to a wrong commit